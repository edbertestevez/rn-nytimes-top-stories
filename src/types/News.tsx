export interface IMultimedia {
  url: string;
  format: string;
  caption: string;
}

export type News = {
  title: string;
  abstract: string;
  url: string;
  uri: string;
  byline: string;
  published_date: Date;
  multimedia: Array<IMultimedia>;
};
