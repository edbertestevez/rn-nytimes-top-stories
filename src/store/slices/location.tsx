import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {ISelectItem} from '../../types/Input';

export interface ILocationState {
  countries: Array<ISelectItem>;
}

export const defaultCountry = [
  {
    label: 'Location',
    value: '',
  },
];

export const initialState: ILocationState = {
  countries: defaultCountry,
};

interface IPayload {
  countries: Array<ISelectItem>;
}

export const locationSlice = createSlice({
  name: 'location',
  initialState,
  reducers: {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    reset: (state) => {
      state = initialState;
    },
    setCountries: (state: ILocationState, action: PayloadAction<IPayload>) => {
      state.countries = [...defaultCountry, ...action.payload.countries];
    },
  },
});

export const locationSliceActions = {
  ...locationSlice.actions,
};
