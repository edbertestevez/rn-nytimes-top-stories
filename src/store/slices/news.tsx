import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {sectionKeys} from '../../constants/sections';
import {News} from '../../types/News';

interface INews {
  list: Array<News>;
  sectionFilter: string;
  keywordFilter: string;
}

const initialState: INews = {
  list: [],
  sectionFilter: sectionKeys.world,
  keywordFilter: '',
};

export const newsSlice = createSlice({
  name: 'news',
  initialState,
  reducers: {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    reset: (state) => {
      state = initialState;
    },
    setList: (state, action: PayloadAction<Array<News>>) => {
      state.list = action.payload;
    },
    setSectionFilter: (state, action: PayloadAction<string>) => {
      state.sectionFilter = action.payload;
    },
    setKeywordFilter: (state, action: PayloadAction<string>) => {
      state.keywordFilter = action.payload;
    },
  },
});

export const newsSliceActions = {
  ...newsSlice.actions,
};
