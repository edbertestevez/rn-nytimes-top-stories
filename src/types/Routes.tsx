import {AppRoutes} from '../navigation/AppRoutes';
import {News} from './News';

//Stack Interfaces
export type NewsStackParams = {
  [AppRoutes.NEWSFEED]: undefined;
  [AppRoutes.NEWS_DETAILS]: {details: News};
};
