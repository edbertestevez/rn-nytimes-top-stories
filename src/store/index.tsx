import AsyncStorage from '@react-native-async-storage/async-storage';
import {
  configureStore,
  getDefaultMiddleware,
  combineReducers,
} from '@reduxjs/toolkit';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import {reducer as network} from 'react-native-offline';

//Slices
import {newsSlice} from './slices/news';
import {locationSlice} from './slices/location';

const reducers = combineReducers({
  news: newsSlice.reducer,
  location: locationSlice.reducer,
  network,
});

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
};

const persistedReducer = persistReducer(persistConfig, reducers);

const store = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware({
    immutableCheck: false,
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
});

const persistor = persistStore(store);

export {store, persistor};

export type AppDispatch = typeof store.dispatch;

export type RootState = ReturnType<typeof store.getState>;
