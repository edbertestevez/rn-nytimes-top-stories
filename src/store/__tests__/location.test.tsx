import configureMockStore from 'redux-mock-store';
import {ISelectItem} from '../../types/Input';
import {initialState, locationSliceActions} from '../slices/location';
import thunk from 'redux-thunk';
import {AppDispatch} from '../../store/index';

const middlewares: any = [thunk];
const mockStore = configureMockStore(middlewares);

describe('[Redux] Location slice testing', () => {
  let store: any;
  let dispatch: AppDispatch;

  beforeEach(() => {
    store = mockStore(initialState);
    store.clearActions();
    dispatch = store.dispatch;
  });

  it('check default countries state', () => {
    let countries = store.getState().countries;
    const initialCountry = countries.find(
      (c: ISelectItem) => c.label === 'Location',
    );

    expect(initialCountry).not.toBeNull();
    expect(initialCountry?.value).toEqual('');
  });

  it('creates <location>/reset when dispatch is successful', () => {
    dispatch(locationSliceActions.reset());

    expect(store.getActions()).toEqual([
      {
        payload: undefined,
        type: 'location/reset',
      },
    ]);

    expect(store.getActions()).toMatchSnapshot();
  });

  it('creates <location>/setCountries when dispatch is successful', () => {
    const testCountries: Array<ISelectItem> = [
      {value: 'country1', label: 'Country 1'},
      {value: 'country2', label: 'Country 2'},
    ];

    dispatch(locationSliceActions.setCountries({countries: testCountries}));

    expect(store.getActions()).toEqual([
      {
        payload: {countries: testCountries},
        type: 'location/setCountries',
      },
    ]);

    expect(store.getActions()).toMatchSnapshot();
  });
});
