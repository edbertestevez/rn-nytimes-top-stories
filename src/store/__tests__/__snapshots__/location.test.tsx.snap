// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Redux] Location slice testing creates <location>/reset when dispatch is successful 1`] = `
Array [
  Object {
    "payload": undefined,
    "type": "location/reset",
  },
]
`;

exports[`[Redux] Location slice testing creates <location>/setCountries when dispatch is successful 1`] = `
Array [
  Object {
    "payload": Object {
      "countries": Array [
        Object {
          "label": "Country 1",
          "value": "country1",
        },
        Object {
          "label": "Country 2",
          "value": "country2",
        },
      ],
    },
    "type": "location/setCountries",
  },
]
`;
